python/ray/tune/execution/tune_controller.py

```diff
@@ -345,6 +347,57 @@ def checkpoint_exists(cls, directory: str) -> bool:
 
         return _experiment_checkpoint_exists(directory)
 
+    def _get_save_fn(self):
+        fs = self._storage.storage_filesystem
+        experiment_fs_path = self._storage.experiment_fs_path
+        exp_state_fs_path = os.path.join(
+            experiment_fs_path, self.experiment_state_file_name
+        )
+        search_alg_filename = self._search_alg.CKPT_FILE_TMPL.format(self._session_str)
+        search_alg_state_fs_path = os.path.join(experiment_fs_path, search_alg_filename)
+        callback_filename = self._callbacks.CKPT_FILE_TMPL.format(self._session_str)
+        callback_state_fs_path = os.path.join(experiment_fs_path, callback_filename)
+
+        # Get state from trial executor and runner
+        # NOTE: This context manager is for Datasets captured in a trial config.
+        # This is the case when *tuning over datasets*.
+        # If the datasets have already been full executed, then serializing
+        # block refs means that this checkpoint is not usable in a new Ray cluster.
+        # This context will serialize the dataset execution plan instead, if available.
+        with out_of_band_serialize_dataset():
+            runner_state = {
+                # Trials
+                "trial_data": list(self._get_trial_checkpoints().values()),
+                # Experiment data
+                "runner_data": self.__getstate__(),
+                # Metadata
+                "stats": {
+                    "start_time": self._start_time,
+                    "timestamp": self._last_checkpoint_time,
+                },
+                "search_alg_filename": search_alg_filename,
+                "callback_filename": callback_filename,
+            }
+
+        search_alg_state = self._search_alg.get_state()
+        callback_state = self._callbacks.get_state()
+
+        def save_fn():
+            with fs.open_output_stream(exp_state_fs_path) as f:
+                f.write(
+                    json.dumps(runner_state, indent=2, cls=TuneFunctionEncoder).encode(
+                        "utf-8"
+                    )
+                )
+
+            with fs.open_output_stream(search_alg_state_fs_path) as f:
+                f.write(ray_pickle.dumps(search_alg_state))
+
+            with fs.open_output_stream(callback_state_fs_path) as f:
+                f.write(ray_pickle.dumps(callback_state))
+
+        return save_fn
```

Comment:
change 1: experiment, searcher, callback state gets uploaded directly from memory

behavioral change: the driver now keeps 1 more copy of all 3 of these things in memory, while the background thread uploads to storage